### PR TITLE
Disable isStartOfType's lookahead when called from isStartOfParameter

### LINKED ISSUE #18123: TypeScript 2.5.1 RC parse error (not in TS 2.4.2)
As of v1.6 of Prettier we are using the TypeScript 2.5 RC and we have had a bug report for some code which worked fine pre-1.6 but is now broken.

Using `tsc` directly, I can confirm that when the following code is parsed with TypeScript 2.4.2 it produces an AST just fine, but when parsed with TypeScript 2.5.1 it produces `"'=' expected."`

**Code**

```ts
const foo = (x:string):void => (
  bar(
    x,
    () => {},
    () => {}
  )
);
```

Original issue: https://github.com/prettier/prettier/issues/2702

### PR DESCRIPTION
**Pull request by sandersn**
`isStartOfType` has lookahead for parenthesized types. Now that unified JSDoc parsing allows types to start parameters, this lookahead compounds with speculative arrow-function parsing's lookahead (which also cares about parentheses). Disable the lookahead when `isStartOfType` is called from `isStartOfParameter`, which was introduced with the addition of unified JSDoc parsing.

Fixes #18123

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -2698,7 +2699,7 @@ namespace ts {
             }
         }
 
-        function isStartOfType(): boolean {
+        function isStartOfType(disableLookahead?: boolean): boolean {
```
Comment (mhegazy) [inline src/compiler/parser.ts]: can we have a more declarative name, like inJSDocType or something?

File: tests/baselines/reference/fatarrowfunctionsOptionalArgsErrors2.errors.txt [hunk 2]
```diff
@@ -18,16 +21,22 @@ tests/cases/compiler/fatarrowfunctionsOptionalArgsErrors2.ts(4,17): error TS1005
 tests/cases/compiler/fatarrowfunctionsOptionalArgsErrors2.ts(4,20): error TS2304: Cannot find name 'a'.
 
 
-==== tests/cases/compiler/fatarrowfunctionsOptionalArgsErrors2.ts (18 errors) ====
+==== tests/cases/compiler/fatarrowfunctionsOptionalArgsErrors2.ts (21 errors) ====
     var tt1 = (a, (b, c)) => a+b+c;
-                  ~
-!!! error TS1003: Identifier expected.
```
Comment (mhegazy) [inline tests/baselines/reference/fatarrowfunctionsOptionalArgsErrors2.errors.txt]: hmm. so we had this test originally? why did not we catch this the first time around then?

Comment (sandersn) [reply tests/baselines/reference/fatarrowfunctionsOptionalArgsErrors2.errors.txt]: I noticed the baseline change but (1) I thought that removal of 3 errors was actually an improvement (2) didn't think it was indicative on any further problem. I also looked at the emitted code and thought it was better -- closer to what was written at least.

File: src/compiler/parser.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/parser.ts]: sure. done.

File: tests/baselines/reference/fatarrowfunctionsOptionalArgsErrors2.errors.txt (hunk 2 above)
Comment (mhegazy) [reply tests/baselines/reference/fatarrowfunctionsOptionalArgsErrors2.errors.txt]: well.. the new error is a parse error. so a bit worse than the 3 semantic ones.

**Comment by JamesHenry** [comment]
Thanks guys 👍

**Comment by mhegazy** [comment]
@sandersn can you port this to release-2.5 as well

**Comment by sandersn** [comment]
It's in release-2.5 now.

**Pull request merged**